python/taichi/lang/impl.py

```diff
@@ -191,6 +191,8 @@ def __getattribute__(self, item):
 def var(dt, shape=None, needs_grad=False):
   if isinstance(shape, numbers.Number):
     shape = (shape,)
+  
+  assert not get_runtime().materialized, "All variables should be declared before materialization"
```

Comment:
Was `import taichi as ti` and `ti.get_runtime()` here necessary?

Reply:
By the way,
`All variables should be declared before materialization`
`All variables should be declared before kernel invocation`
`All variables should be declared before kernel invocation or assignment`
`All variables should be declared before any variable being used`
Which message is the best?

Reply:
Awesome! 

> Was import taichi as ti and ti.get_runtime() here necessary?

Simply `get_runtime()` should work.

I would suggest ``No new variables can be declared after kernel invocations or Python-scope tensor accesses.`` since it's more direct.